pkg/gitops/filter.go

```diff
@@ -0,0 +1,80 @@
+package gitops
+
+import (
+    "fmt"
+	"time"
+    "strings"
+    "path/filepath"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/object"
+)
+
+func (cr *CustomRepo) FilterCommits(author *string, since *time.Time, until *time.Time, resource *string, namespace *string, name *string) ([]object.Commit, error) {
+	var logopts git.LogOptions
+	var filteredCommits []object.Commit
+
+	cr.Mutex.Lock()
+	defer cr.Mutex.Unlock()
+	ref, err := cr.Repo.Head()
+	if err != nil {
+		return filteredCommits, fmt.Errorf("could not get ref head from repository: %w", err)
+	}
+
+	logopts.From = ref.Hash()
+	if !since.IsZero() && since != nil {
+		logopts.Since = since
+	}
+	if !since.IsZero() && until != nil {
+		logopts.Until = until
+	}
+
+	if *resource != "" && *namespace == "" && *name != "" {
+        return filteredCommits, fmt.Errorf("error (FilterCommits): cannot provide a resource without a namespace")
+    } else if *resource == "" && *namespace != "" {
+        resources := []string{"k8s-policies", "antrea-policies", "antrea-cluster-policies", "antrea-tiers"}
+        for _, r := range resources {
+            logopts.PathFilter = func(path string) bool {
+                fpath := filepath.Join(r, *namespace, *name)
+                if strings.Contains(path, fpath) {
+                    return true
+                }
+                return false
+            }
+            tempCommits, err := cr.filter(author, logopts)
+            if err != nil {
+                return filteredCommits, err
+            }
+            for _, tc := range tempCommits {
+                filteredCommits = append(filteredCommits, tc)
+            }
+        }
+	} else {
+        fpath := filepath.Join(*resource, *namespace, *name)
+        if fpath != "" {
+            logopts.PathFilter = func(path string) bool {
+                if strings.Contains(path, fpath) {
+                    return true
+                }
+                return false
+            }
+        }
+        filteredCommits, err = cr.filter(author, logopts)
+    }
+    return filteredCommits, err
+}
+
+func (cr *CustomRepo) filter(author *string, logopts git.LogOptions) ([]object.Commit, error) {
```

Comment:
can we make this function a bit more generic so that it's more future-proof. Something like this:

```go
type customFilterFn func(*object.Commit) bool

func (cr *CustomRepo) filter(logopts *git.LogOptions, customFilters ...customFilterFn) ([]*object.Commit, error) {
    // get the commits from go-git using logopts, then apply all the custom filters on the result
}
```